File: 0632-smallest-range-covering-elements-from-k-lists/0632-smallest-range-covering-elements-from-k-lists.py
class Solution:
    def smallestRange(self, nums: List[List[int]]) -> List[int]:
        heap = []
        max_value = 0
        
        for i in range(len(nums)):
            heappush(heap, [nums[i][0], i, 0])
            max_value = max(max_value, nums[i][0])
            
        ans = [heap[0][0], max_value]
        
        while True:
            _, row, col = heappop(heap)
            if col == len(nums[row]) - 1:
                break
            
            next_num = nums[row][col + 1]
            heappush(heap, [next_num, row, col + 1])
            max_value = max(max_value, next_num)
            
            if max_value - heap[0][0] < ans[1] - ans[0]:
                ans = [heap[0][0], max_value]
        
        return ans

File: 0632-smallest-range-covering-elements-from-k-lists/0632-smallest-range-covering-elements-from-k-lists.cpp
class Solution {
    struct compare{
        bool operator()(pair<int, pair<int, int>>& a, pair<int, pair<int, int>>& b){
            if (a.first == b.first){
                if (a.second.first == b.second.first) return a.second.second > b.second.second;
                return a.second.first > b.second.first;
            }
            return a.first > b.first;
        }
    };
    
public:
    vector<int> smallestRange(vector<vector<int>>& nums) {
        priority_queue<pair<int, pair<int, int>>, vector<pair<int, pair<int, int>>>, compare> pq;
        int maxVal = 0;
        for (int i = 0; i < nums.size(); ++i){
            pq.push({nums[i][0], {i, 0}});
            maxVal = max(maxVal, nums[i][0]);
        }
        vector<int> ans{pq.top().first, maxVal};
        
        while(true){
            auto [_, mat] = pq.top();
            auto [row, col] = mat;
            pq.pop();
            
            if (col == nums[row].size() - 1) break;
            int nextNum = nums[row][col + 1];
            pq.push({nextNum, {row, col + 1}});
            maxVal = max(maxVal, nextNum);
            if (maxVal - pq.top().first < ans[1] - ans[0])
                ans = {pq.top().first, maxVal};
        }
        
        return ans;
    }
};